src/btree/bt_npos.c

```diff
@@ -0,0 +1,352 @@
+ *    === Calculating a page's normalized position.
+ *
+ * As opposed to finding a page from its normalized position, the process goes back from the page
+ * up to the root. The process is a reverse of the finding process.
+ *
+ * A nuance is that because there is a whole range of numbers corresponding to a page, the user
+ * can choose a starting position within the page. Say, numbers closer to 0 will point to somewhere
+ * closer to the beginning of the page, and numbers closer to 1 will point close to the end.
+ *
+ * If there's only one level, then the normalized position is just the page's index (plus fractional
+ * starting point) divided by the number of pages at the parent level.
+ *
+ * If there are multiple levels, the process is repeated until we reach the root with starting
+ * point being whatever has been calculated at the previous level.
+ *
+ * NOTE that starting points 0 and 1 are corner cases and can lead you to an adjacent page when
+ * retrieving a page because of rounding errors.
+ * To reliably get back to the same page, the best starting point is 0.5.
+ *
+ * A useful side effect is that using starting numbers outside of 0 .. 1 range will lead you to
+ * adjacent pages. This can be used to iterate over pages without storing any hazard pointers.
+ * An example of this can be found in test.
+ *
+ *   === Precision considerations.
+ *
+ * Because the precision of the position if affected by tree's structure, it can be used to quantify
+ * the shape of the tree. The integral difference of all page's normalized positions and their
+ * actual positions can be used to estimate the tree's quality.
+ *
+ * Note that the tree shape in memory can significantly diverge from the tree shape on disk.
+ *
+ *   === How many pages can be addressed by a double precision number?
+ *
+ * The maximum number of pages that can be addressed by a double is roughly 2^53 = ~ 10^16
+ * (where 53 is the number of bits in a double mantissa).
+ * We have multiple orders of magnitude spare by now.
+ *
+ * For distributed storage it still can be not enough (the dataset size can exceed
+ * petabytes or exabytes), then we can shift to using 64-bit or 128-bit fixed-point numbers.
+ *
+ */
+
+/*
+ * !!!
+ * __wt_page_npos --
+ *     Get the page's normalized position in the tree.
+ *     - If `path_str_offsetp` is set, return a string representation of the page's path.
+ *     - `start` is a position within the leaf page: 0 .. 1.
+ *       * When calculating a leaf page's position, use 0.5 to get the middle of the page.
+ *       * 0 and 1 are corner cases and can lead you to an adjacent page.
+ *       * Numbers outside of 0 .. 1 range will lead you to a prev/next page.
+ */
+double
+__wt_page_npos(WT_SESSION_IMPL *session, WT_REF *ref, double start, char *path_str,
+  size_t *path_str_offsetp, size_t path_str_sz_max)
+{
+    WT_PAGE_INDEX *pindex;
+    double npos;
+    uint32_t entries, slot;
+    int unused = 1; /* WT_UNUSED(snprintf) is cooked in GCC */
```

Comment:
np: Fullstop at end of comments in WT, including the below "in case of error" comment. There are probably other too in this PR but I won't nitpick all of them.

Reply:
Done all